src/cdk/testing/component-harness.ts

```diff
@@ -76,72 +111,88 @@ export interface LocatorFactory {
+   * Creates an asynchronous locator function that can be used to find a `ComponentHarness` instance
+   * or element under the root element of this `LocatorFactory`.
+   * @param {...*} queries A list of queries specifying which harnesses and elements to search for:
+   *   - A `string` searches for elements matching the selector specified by the string.
+   *   - A `ComponentHarness` constructor searches for `ComponentHarness` instances matching the
+   *     given class.
+   *   - A `HarnessPredicate` searches for `ComponentHarness` instances matching the given
+   *     predicate.
+   * @return An asynchronous locator function that searches for and returns a `Promise` for the
+   *   first element or harness matching the given search criteria. Matches are ordered first by
+   *   order in the DOM, and second by order in the queries list. If no matches are found, the
+   *   `Promise` rejects.
+   *
+   * e.g. Given the following DOM: `<div id="d1" /><div id="d2" />`, and assuming
+   * `DivHarness.hostSelector === 'div'`:
+   * - `await lf.locatorFor(DivHarness, 'div')()` gets a `DivHarness` instance for `#d1`
+   * - `await lf.locatorFor('div', DivHarness)()` gets a `TestElement` instance for `#d1`
+   * - `await lf.locatorFor('span')()` throws because the `Promise` rejects.
    */
-  locatorForOptional<T extends ComponentHarness>(
-      harnessType: ComponentHarnessConstructor<T> | HarnessPredicate<T>): AsyncFactoryFn<T | null>;
+  locatorFor<T extends (HarnessQuery<any> | string)[]>(...queries: T):
+      AsyncFactoryFn<LocatorFnResult<T>>;
 
   /**
-   * Creates an asynchronous locator function that can be used to search for a list of elements with
-   * the given selector under the root element of this `LocatorFactory`. When the resulting locator
-   * function is invoked, a list of matching elements is returned.
-   * @param selector The selector for the element that the locator function should search for.
-   * @return An asynchronous locator function that searches for elements with the given selector,
-   *     and either finds one or throws an error
+   * Creates an asynchronous locator function that can be used to find a `ComponentHarness` instance
+   * or element under the root element of this `LocatorFactory`.
+   * @param {...*} queries A list of queries specifying which harnesses and elements to search for:
+   *   - A `string` searches for elements matching the selector specified by the string.
+   *   - A `ComponentHarness` constructor searches for `ComponentHarness` instances matching the
+   *     given class.
+   *   - A `HarnessPredicate` searches for `ComponentHarness` instances matching the given
+   *     predicate.
+   * @return An asynchronous locator function that searches for and returns a `Promise` for the
+   *   first element or harness matching the given search criteria. Matches are ordered first by
+   *   order in the DOM, and second by order in the queries list. If no matches are found, the
```

Comment:
What's your reasoning behind ordering them this way instead of always being in DOM order irrelevant of the query order?